### PR TITLE
Use regex to detect exclusions instead of string.Contains

### PR DESCRIPTION
**Pull request by jonesbrianc26**
I added a struct which uses a regex to determine whether or not a file should be excluded.

if .git is excluded, it excludes .git and .git/*, and will not exclude .gitignore

in general if you have a list of exclusions `.git, .gitignore` the generated regexp will look like:
`^\.git$|^\.git/.*$|^\.gitignore$|^\.gitignore/.*$` 

Fixes: https://github.com/kshvmdn/fsql/issues/5

### CODE REVIEW

**Comment by kshvmdn** [comment]
@jonesbrianc26 Thanks for the contribution! I'll have to review/test it in more depth tomorrow, but I took a brief glance and it's looking very nice so far.

**Comment by jonesbrianc26** [comment]
My test additions are pretty minimal, but I think this would be a perfect chance to add something like mentioned here https://github.com/kshvmdn/fsql/issues/20

**Comment by kshvmdn** [comment]
I've actually gotten started on something for that.

The temporary directory idea that was mentioned in #20 sounds like a cool idea, but might be a little excessive for this. Go [automatically ignores anything inside a `testdata` directory](https://golang.org/cmd/go/#hdr-Description_of_package_lists), so the idea is to fill that directory with arbitrary files/fixtures and then use those in the tests.

**Comment by jonesbrianc26** [comment]
I'm not too familiar with go tests unfortunately, but a good mock library would he helpful, but data driven tests would definitely fit the case here I think 👍

File: query/excluder.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package query
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+//Excluder allows us to support different methods of excluding in the future
```
Comment (kshvmdn) [inline query/excluder.go]: ✔️

I think we can use this same interface to detect paths we've seen and apply regex to that to check for leading/trailing slashes (this is also naively implemented with a simple map and no regex right now).

Let me know if you'd like to work on that as well (should be in a different PR).

File: query/excluder.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	recursive  bool
+	exclusions []string
+	regex      string
+}
+
+//ShouldExclude will return a boolean denoting whether or not the path should be excluded based on the given slice of exclusions
+func (r *RegexpExclude) ShouldExclude(path string) bool {
+	var b bool
+	var ok error
+	if r.regex == "" {
```
Comment (kshvmdn) [inline query/excluder.go]: When no exclusion is provided, this is running every single time and nothing is being outputted (since everything matches `""`).

The solution here would be to track if the expression has been built or not, and if it's been built and still is empty, then this method should return false without even running `regexp.MatchString`.

File: query/excluder.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+//Excluder allows us to support different methods of excluding in the future
+type Excluder interface {
+	ShouldExclude(path string) bool
+}
+
+//RegexpExclude uses regular expressions to tell if a file/path should be excluded
+type RegexpExclude struct {
+	recursive  bool
```
Comment (kshvmdn) [inline query/excluder.go]: Don't think this is being used anywhere? Would be a good idea to remove it (interesting in hearing what this was intended for though).

File: query/excluder.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	exclusions []string
+	regex      string
```
Comment (kshvmdn) [inline query/excluder.go]: I think it'd be more appropriate if this was of type [`*regexp.Regexp`](https://golang.org/pkg/regexp/#Regexp).

File: query/excluder.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+				or(mustEnd(mustBegin(escape(exclusion)+"/.*")),
+					mustEnd(mustBegin(escape(exclusion[:len(exclusion)-1]))))
+		} else {
+			curr =
+				or(mustEnd(mustBegin(escape(exclusion))),
+					mustEnd(mustBegin(escape(exclusion)+"/.*")))
+		}
+		regex = or(prev, curr)
+	}
+	r.regex = regex
```
Comment (kshvmdn) [inline query/excluder.go]: Since `r.regex` is of type `*regexp.Regexp` now, this should read

```go
r.regex = regexp.MustCompile(regex)
```

Note that this panics on invalid regex, which I think is the appropriate solution here, since it means the input was invalid/malformed.

File: query/excluder.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+//ShouldExclude will return a boolean denoting whether or not the path should be excluded based on the given slice of exclusions
+func (r *RegexpExclude) ShouldExclude(path string) bool {
+	var b bool
+	var ok error
+	if r.regex == "" {
+		r.buildRegex()
+	}
+
+	if b, ok = regexp.MatchString(r.regex, path); not(ok) {
```
Comment (kshvmdn) [inline query/excluder.go]: This can now be reduced to:

```go
return r.regex.MatchString(path)
```

File: query/query_test.go [hunk 7]
```diff
@@ -1 +1,41 @@
 package query
+
+import "testing"
+
+func TestShouldExclude_ExpectAllExcluded(t *testing.T) {
+	exclusions := make([]string, 0)
+	exclusions = append(exclusions, ".git", ".gitignore")
```
Comment (kshvmdn) [inline query/query_test.go]: A little bit nitpicky, but you can do the equivalent in a single line with:

```go
exclusions := []string{".git", ".gitignore"}
```

File: query/query_test.go [hunk 8]
```diff
@@ -1 +1,41 @@
...
+
+	b = excluder.ShouldExclude(".gitignore")
+	if b == false {
+		t.Fail()
+	}
+}
+
+func TestShouldExclude_ExpectNotExcluded(t *testing.T) {
+	exclusions := make([]string, 0)
+	exclusions = append(exclusions, ".git")
```
Comment (kshvmdn) [inline query/query_test.go]: Here too 😄

File: query/excluder.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	return b
+}
+
+func (r *RegexpExclude) buildRegex() {
+	var regex string
+	var prev string
+	var curr string
+	for _, exclusion := range r.exclusions {
+		prev = curr
+		if strings.HasSuffix(exclusion, "/") {
```
Comment (kshvmdn) [inline query/excluder.go]: I like what's happening here, but I think there's a simpler way to do it. So instead of `^.git$|^.git/.*$`,  we can use `^.git(/.*)?$`.

You can simplify this loop by first [right-trimming](https://golang.org/pkg/strings/#TrimRight) any `/` and [replacing](https://golang.org/pkg/strings/#Replace) all `.` with `\\.`, then wrapping the result with `^` and `(/.*)?$`.

I suggest creating a slice where each exclusion is put into that form, and then joining that slice with [`strings.Join`](https://golang.org/pkg/strings/#Join) after the loop is done.

File: query/excluder.go (hunk 3 above)
Comment (jonesbrianc26) [inline query/excluder.go]: Was toying with the idea of preemptively excluding paths here, but decided against that for the scope of this PR. The idea being you could pass in flags to determine the behavior of the struct

File: query/excluder.go (hunk 9 above)
Comment (jonesbrianc26) [inline query/excluder.go]: I like his alot.  Definitely simplifies the logic

File: query/excluder.go (hunk 5 above)
Comment (jonesbrianc26) [inline query/excluder.go]: I agree. Behavior now is to print out the same broken error message, so I agree a panic would be preferred

**Comment by jonesbrianc26** [comment]
@kshvmdn

Refactored a bit of the code.

I went with what you said for all of these, great catches!

File: query/excluder.go (hunk 1 above)
Comment (jonesbrianc26) [inline query/excluder.go]: I can make an issue for this, see if anyone picks it up, otherwise I'll gladly take care of it!

File: query/excluder.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	exclusions []string
+	regex      *regexp.Regexp
+}
+
+//ShouldExclude will return a boolean denoting whether or not the path should be excluded based on the given slice of exclusions
+func (r *RegexpExclude) ShouldExclude(path string) bool {
+	if r.regex == nil {
+		r.buildRegex()
+	}
+	if r.regex.String() == "" {
```
Comment (kshvmdn) [inline query/excluder.go]: nice!

**Comment by kshvmdn** [comment]
Thanks @jonesbrianc26!
